### DISCUSSION TITLE
Uploading files from SvelteKit endpoints

### DISCUSSION THREAD

**Discussion by stephane-vanraes** [pocketbase/pocketbase#278] [Q&A] [answered]
I am working on an application that allows the user to upload images, these images have to pass through a preprocessing step before storing them in the database.  For this I have setup an api in node and use a regular form submit client side to pass data to this endpoint.

Now I am however struggling to send the data further to Pocketbase itself.

I tried the following very basic setup to just pass on the `FormData` object: 
```js
const formData = await request.formData();
const record = await db.records.create('images', formData)
```

But I get the following error:
> Failed to load the submitted data due to invalid formatting.

Picking off some of the fields individually works fine: 
```
const formData = await request.formData();
const record = await db.records.create('images', {
    alt: formData.get('alt'),
})
```

Any suggestion on how to make file uploads work from the server side ?

**Comment by ganigeorgiev** [comment]
Could you start the PocketBase binary with `--debug` (or check in the "Admin UI > Logs" the more detailed request error)?

**Reply by stephane-vanraes** [reply]
I am getting the following error in the logs:
```
{
  "errorDetails": "Unsupported request Content-Type.",
  "errorMessage": "Failed to load the submitted data due to invalid formatting."
}
```

The upload itself is done with a plain form like
```
<form method="POST" action="/images" >
	<input type="file" name="file" />
	<input type="text" name="alt" value="my placeholder" />
	<button type="submit">Submit</button>
</form>
```
Which _afaik_ should send it as a multipart content type

**Reply by ganigeorgiev** [reply]
Hm, it seems that it is not recognized as `multipart/form-data` body (or the multipart/form-data boundary is invalid).

Could you confirm that `formData` is actually a [`FormData`](https://developer.mozilla.org/en-US/docs/Web/API/FormData) instance? Internally the JavaScript SDK uses `fetch()` so every body parameter that fetch supports should be supported by the SDK as well.

Alternatively, just for test you could also try constructing your own FormData object from the current one:
```js
const createFormData = new FormData()
createFormData.append("file", formData.get("file"));
createFormData.append("alt", formData.get("alt"));

await db.records.create("images", createFormData)
```

**Reply by stephane-vanraes** [reply]
`formData` is of type `FormData`, so that is ok.

Creating a new FormData object is a bit problematic as the type does not exist in Node, using the npm package `form-data` gives an error on that one `source.on is not a function`.

I will have a further look, thanks for the input 👍

**Reply by ganigeorgiev** [reply]
Which version of Node are you using? Node 18 supports it and has a global FormData.

**Reply by ganigeorgiev** [reply]
For older Node versions you could use a polyfill (if you are not using already) and use their `FormData` object. Here is an example using [`node-fetch`](https://www.npmjs.com/package/node-fetch):

```js
import { FormData } from 'node-fetch';
```

--- 

Regarding the `source.on is not a function` error - this is probably because the assigned value is not a `string`, `Buffer` or `Stream` (eg. if you load the file from the localfilesytem you could use something like `fs.createReadStream("/path/to/your/file.png")`.

**Reply by stephane-vanraes** [reply]
Thank you for the ideas.
I update node to v18 but sadly enough did not get it working
Moving the code to the client and pushing directly there to the PocketBase seems to work though, so the difference might be in a different implementation of  `FormData` in node vs the browser.
Going to try more today and will keep you updated

**Comment by stephane-vanraes** [comment]
@ganigeorgiev I think I have found _something_
In the pocketbase code itself I added in the  `send` function the following:

```js
 console.log(e)
 e.headers = Object.assign({}, e.headers, {
   "Content-Type": "multipart/form-data",
})
console.log(e)
```

The first log shows that `e` does not have headers, which I believe causes issues down the line where the headers are copied to be used in the GET request fired by pocketbase itself. (The second log is just to verify the headers have been set...)

With this addition I get the following error:
> 2022/08/09 10:32:09 no multipart boundary param in Content-Type

If I blindly copy the headers that the browser sends to my API and put them in the code like this:
```js
"Content-Type": "multipart/form-data; boundary=----WebKitFormBoundaryAleFysfncMKAqhHz",
``` 
I get 
> 2022/08/09 10:34:31 multipart: NextPart: EOF

I am starting to believe this brings us a touch closer to solving this issue as it is starting to look like the headers are missing when doing the call in node/server side.

For reference the current code used is
```js
const f = new FormData();
f.append('file', formData.get('file'));
f.append('alt', formData.get('alt'));

const record = await db.records.create('images', f);
```

**Reply by ganigeorgiev** [reply]
@stephane-vanraes when you use a FormData as body to fetch you don't need to specify the Content-type header, by spec it will set it for you since the boundary will be different with each new request. I've tested the following in node18:

```js
import fs from "fs";
import PocketBase from "pocketbase";

const client = new PocketBase("http://127.0.0.1:8090");

await client.admins.authViaEmail("test@example.com", "1234567890");

const formData = new FormData();
formData.append("alt", "Hello world!");
formData.append("file", new Blob([fs.readFileSync("./test.png")]), "test.png");

await client.records.create("posts", formData);
```

It turns out that the file need to be a blob, it cannot be a stream. I was confused by the custom [`form-data/form-data`](https://github.com/form-data/form-data) library where their example shows that the file could be loaded from a stream but this doesn't seem to work with the native fetch.

**Reply by ganigeorgiev** [reply]
Or in other words - in your example you could try to log `formData.get('file')` to see what its type and whether it is a Blob (or File, which is a special type of Blob):
```js
console.log(formData.get("file") instanceof Blob)
```

**Reply by stephane-vanraes** [reply]
Something else must be wrong, using the code directly in a plain script works perfect and `formData.get('file')` is of type `File` in my endpoint. Wondering now if the framework (SvelteKit) I am using is somehow messing up the requests somewhere.

Looks like this is not a PocketBase issue...

**Reply by stephane-vanraes** [reply]
That's it!
Using `node-fetch` to override the global `fetch` available to me in SvelteKit and calling the API endpoint the 'old fashioned way': 
```
 const formData = await request.formData();
 await fetch(API_URL + '/api/collections/images/records', {
    method: 'POST',
    body: formData,
 });
```` 

Works perfectly fine (except that it means I have to manually add authentication headers here, will figure that out next).
Seems like a bug in SvelteKit, will go make an issue over there 💯

**Reply by ganigeorgiev** [reply]
If you load node-fetch globally before calling the PocketBase client you should be able to use the SDK methods (the SDK fetch call is not prefixed and it uses whatever available global `fetch` function there is, aka. either `window.fetch` or `global.fetch`).

**Comment by stephane-vanraes** [comment, marked as answer]
The issue seems to be with SvelteKit and more precisely [undici](https://www.npmjs.com/package/undici) which has an incomplete feature set for `fetch`.
A workaround for this is to simply import `node-fetch` where required to override undici for the time being.

**Reply by stephane-vanraes** [reply]
Likely this issue: https://github.com/nodejs/undici/issues/974
